system/lib/emmalloc.cpp

```diff
@@ -0,0 +1,1187 @@
+  return 31 - __builtin_clz(x);
+}
+
+// Constants
+
+// All allocations are aligned to this value.
+static const size_t ALIGNMENT = 8;
+
+// Even allocating 1 byte incurs this much actual payload
+// allocation. This is our minimum bin size.
+static const size_t ALLOC_UNIT = ALIGNMENT;
+
+// How big the metadata is in each region. It is convenient
+// that this is identical to the above values.
+static const size_t METADATA_SIZE = ALLOC_UNIT;
+
+// How big a minimal region is.
+static const size_t MIN_REGION_SIZE = METADATA_SIZE + ALLOC_UNIT;
+
+// Constant utilities
+
+// Align a pointer, increasing it upwards as necessary
+static size_t alignUp(size_t ptr) {
+  return (size_t(ptr) + ALIGNMENT - 1) & -ALIGNMENT;
+}
+
+static void* alignUpPointer(void* ptr) {
+  return (void*)alignUp(size_t(ptr));
+}
+
+//
+// Data structures
+//
+
+struct Region;
+
+// Information memory that is a free list, i.e., may
+// be reused.
+// Note how this can fit instead of the payload (as
+// the payload is a multiple of MIN_ALLOC).
+struct FreeInfo {
+  // free lists are doubly-linked lists
+  FreeInfo* _prev;
+  FreeInfo* _next;
+
+  FreeInfo*& prev() { return _prev; }
+  FreeInfo*& next() { return _next; }
+};
+
+// The first region of memory.
+static Region* firstRegion = NULL;
+
+// The last region of memory. It's important to know the end
+// since we may append to it.
+static Region* lastRegion = NULL;
+
+// A contiguous region of memory. Metadata at the beginning describes it,
+// after which is the "payload", the sections that user code calling
+// malloc can use.
+struct Region {
+  // Whether this region is in use or not.
+  size_t _used : 1;
+
+  // The total size of the section of memory this is associated
+  // with and contained in.
+  // That includes the metadata itself and the payload memory after,
+  // which includes the used and unused portions of it.
+  // FIXME: Shift by 1, as our size is even anyhow?
+  //        Or, disallow allocation of half the total space or above.
+  //        Browsers barely allow allocating 2^31 anyhow, so inside that
+  //        space we can just allocate something smaller than it.
+  size_t _totalSize : 31;
+
+  // Each memory area knows its previous neighbor, as we hope to merge them.
+  // To compute the next neighbor we can use the total size, and to know
+  // if a neighbor exists we can compare the region to lastRegion
+  Region* _prev;
+
+  // Up to here was the fixed metadata, of size 16. The rest is either
+  // the payload, or freelist info.
+  union {
+    FreeInfo _freeInfo;
+    char _payload[];
+  };
+
+  size_t getTotalSize() { return _totalSize; }
+  void setTotalSize(size_t x) { _totalSize = x; }
+  void incTotalSize(size_t x) { _totalSize += x; }
+  void decTotalSize(size_t x) { _totalSize -= x; }
+
+  size_t getUsed() { return _used; }
+  void setUsed(size_t x) { _used = x; }
+
+  Region*& prev() { return _prev; }
+  // The next region is not, as we compute it on the fly
+  Region* next() {
```

Comment:
Should this function and all the above up to `getTotalSize` be `static`?

Reply:
oh nm, first read that these were global scope functions, but they are instead members of `Region`.